packages/types/src/internal-msg/offscreen.ts

```diff
@@ -10,29 +14,36 @@ export type ActionBuildMessage = InternalMessage<
 
 export type OffscreenMessage = ActionBuildMessage;
 export type OffscreenRequest = InternalRequest<OffscreenMessage>;
+export type OffscreenResponse = InternalResponse<OffscreenMessage>;
 
 export interface ActionBuildRequest {
-  transactionPlan: JsonObject & { actions: JsonValue[] };
-  witness: JsonObject;
+  transactionPlan: Jsonified<TransactionPlan> & {
+    actions: Jsonified<TransactionPlan['actions']>;
+  };
+  witness: Jsonified<WitnessData>;
   fullViewingKey: string;
 }
-export type ActionBuildResponse = ActionJsonValue<ActionCase>[];
+export type ActionBuildResponse = Jsonified<Action>[];
 
 export type WasmBuildActionInput = ActionBuildRequest & { actionPlanIndex: number };
-
-type ActionCase = NonNullable<Action['action']['case']>;
-type ActionJsonValue<C extends ActionCase> = Record<C, JsonValue>;
-
-const hasActionPlanJsonArray = (x: JsonValue): x is JsonObject & { actions: JsonValue[] } =>
-  x != null && typeof x === 'object' && 'actions' in x && Array.isArray(x['actions']);
+export type WasmBuildActionOutput = Jsonified<Action>;
 
 export const isActionBuildRequest = (req: unknown): req is ActionBuildRequest =>
   req != null &&
   typeof req === 'object' &&
   'transactionPlan' in req &&
-  hasActionPlanJsonArray(req.transactionPlan as JsonObject) &&
+  req.transactionPlan != null &&
+  typeof req.transactionPlan === 'object' &&
+  'actions' in req.transactionPlan &&
+  Array.isArray(req.transactionPlan.actions) &&
   'witness' in req &&
-  typeof req.witness === 'object' &&
   req.witness != null &&
+  typeof req.witness === 'object' &&
   'fullViewingKey' in req &&
   typeof req.fullViewingKey === 'string';
+
+export const isWasmBuildActionInput = (req: unknown): req is WasmBuildActionInput =>
+  isActionBuildRequest(req) && 'actionPlanIndex' in req && typeof req.actionPlanIndex === 'number';
+
+export const isWasmBuildActionOutput = (res: unknown): res is WasmBuildActionOutput =>
+  res != null && typeof res === 'object';
```

Comment:
Is this too generic? This basically means any object could be a WasmBuildActionOutput?

Reply:
replaced with a JsonObject type guard. we can't guard more than this without inspecting protojson, and we just parse it with fromJson immediately anyway, so it's fine.